File: python/KRVFL_plus.py
import numpy as np
from sklearn.metrics.pairwise import rbf_kernel, linear_kernel, chi2_kernel, polynomial_kernel, \
                                        additive_chi2_kernel, laplacian_kernel
from sklearn.preprocessing import OneHotEncoder

class KRVFL_plus(object):
    """
    The code of kernel random vector functional link-plus (KRVFL+) network is implemented in Python. If you would like
    to use it in your researches, please cite the paper "Zhang, Peng-Bo, and Yang, Zhi-Xin. A new learning paradigm for
    random vector functional-link network: RVFL+. Neural Networks 122 (2020) pp.94-105".

    Parameters
    ----------
    kernel_name : default='rbf', the kernel function is used in KRVFL+. Please select a kernel function from the below
    set: {'rbf', 'linear', 'add_chi2', 'chi2', 'poly', 'laplace'}, and other kernel function is not supported now. In
         the later, we will add more kernel functions.
    type: default='classification', classification or regression,


    Example
    --------
    model = KRVFL_plus()
    model.fit(train_x, addition_x, train_y)
    y_hat = model.predict(train_x, test_x)
    """

    def __init__(self, kernel_name = 'rbf', type = 'classification'):

        self.kernel_name = kernel_name
        self.type = type
        self.kernel_dict = {"rbf": lambda x, y = None: rbf_kernel(x, y),
                            "linear": lambda x, y = None: linear_kernel(x, y),
                            "add_chi2": lambda x, y = None: additive_chi2_kernel(x, y),
                            "chi2": lambda x, y = None: chi2_kernel(x, y),
                            "poly": lambda x, y = None: polynomial_kernel(x, y),
                            "laplace": lambda x, y = None: laplacian_kernel(x, y)}

    def _transform_label(self, y):
        enc = OneHotEncoder(handle_unknown='ignore')
        try:
            target = enc.fit_transform(y).toarray()
            print('the label can be transformed directly using onehotencoder')
        except:
            target = enc.fit_transform(y.reshape(-1, 1)).toarray()
            print('the label must be reshaped before being transformed')
        return target

    def _softmax(self, x):

        max_value = np.max(x,axis=1,keepdims=True)
        e_x = np.exp(x - max_value) 
        sum_value = np.sum(e_x,axis=1,keepdims=True)
        y = e_x / sum_value
        return y

    def fit(self, train_x, addition_x, train_y, C = 0.1, gamma = 1000):
        """
        Params:
        ---------
        :param train_x: a NumofSamples * NumofFeatures matrix, training data
        :param addition_x: a NumofSamples * NumofPFeatures matrix, the privileged information data
        :param train_y: training label
        :param C: default=0.1, the penelty parameter
        :param gamma: default = 1000, the trade-off parameter balancing between training data and privileged data
        """
        omega1, omega2 = self.kernel_dict["linear"](train_x), self.kernel_dict[self.kernel_name](train_x)
        omega_p_1, omega_p_2 = self.kernel_dict["linear"](addition_x), self.kernel_dict[self.kernel_name](addition_x)
        if self.type == 'classification':
            one_hot_target = self._transform_label(train_y)
        elif self.type == 'regression':
            one_hot_target = train_y
        else:
            raise Exception("The type is not supported now! please select classification or regression.")
        self.beta = np.linalg.inv(omega1 + omega2 + 1/gamma*(omega_p_1 + omega_p_2) + np.eye(train_x.shape[0])/C).\
            dot(one_hot_target - C / gamma * (omega_p_1 + omega_p_2).dot(np.ones(one_hot_target.shape)))


    def predict(self, train_x, test_x):
        """

        :param train_x:  a NumofSamples * NumofFeatures matrix, training data, building the kernel matrix
        :param test_x: a NumofTestSamples * NumofFeatures matrix, test data
        :return: y_hat, the predicted labels
        """
        y_hat = self.predict_proba(train_x, test_x)
        if self.type == 'classification':
            y_hat = np.argmax(y_hat, axis=1)
            return y_hat
        else:
            return y_hat

    def predict_proba(self, train_x, test_x):

        omega_test_1,  omega_test_2 = self.kernel_dict["linear"](test_x, train_x), \
                                      self.kernel_dict[self.kernel_name](test_x, train_x)
        y_hat_temp = (omega_test_1 + omega_test_2).dot(self.beta)
        if self.type == "classification":

            y_hat_prob = self._softmax(y_hat_temp)
            return y_hat_prob
        else:
            return y_hat_temp

File: matlab/KRVFL_plus.m
function [TrainingTime, TestingTime, TrainingAccuracy, TestingAccuracy] = KRVFL_plus(trainX,additionX, trainY,testX,testY, option)

%% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This is an implementation of KRVFL+ with Matlab 2015b. 
% If you use it in your research, please cite the paper "Zhang Peng-Bo, and Yang Zhi-Xin. A new 
% learning paradigm for random vector functional-link network: RVFL+.Neural Networks 122 (2020) pp.94-105".
%% Parameters:
% trainX: Nsamples * Nfeatures matrix
% trainyY: Nsamples * 1 matrix(vector)
% addtionX: Nsamples * N-additional-features matrix
% testX: Msamples * Nfeatures matrix
% testY: Msamples * 1 matrix(vector)
% option: 
%   option.Type: 1 classification 0 regression
%   option.Kernel_class: 'RBF_kernel', 'lin_kernel', 'poly_kernel', 'wav_kernel'
%   option.Kernel_width
%   option.C
%   option.gamma
%% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if ~isfield(option,'Type')|| isempty(option.Type)
    option.Type = 1;
  
end
if ~isfield(option,'Kernel_class')|| isempty(option.Kernel_class)
    option.Kernel_class='RBF_kernel';
   
end
if ~isfield(option,'Kernel_width')|| isempty(option.Kernel_width)
    option.Kernel_width=0.02;
   
end
if ~isfield(option,'C')|| isempty(option.C)
    option.C=0.1;
   
end
if ~isfield(option,'gamma')|| isempty(option.gamma)
    option.gamma=1000;
   
end

%% Load data
gamma = option.gamma;
C = option.C;
Kernel_width = option.Kernel_width;
Kernel_class = option.Kernel_class;

%%%%%%%%%%% Load training dataset
T=trainY';
P=trainX';

%%%%%%%%%%% Load additional information
PF = additionX';

%%%%%%%%%%% Load testing dataset
TV.T = testY';
TV.P = testX';

NumberofTrainingData=size(P,2);
NumberofTestingData=size(TV.P,2);

%% Convert labels into one-hot vectors 
if option.Type == 1
	U_trainY=unique(trainY);
	nclass=numel(U_trainY);
	trainY_temp=zeros(nclass, numel(trainY));
	for i=1:nclass
         idx= trainY==U_trainY(i);
        
         trainY_temp(i, idx)=1;
	end
	T = trainY_temp;
end
%% Training using training data and privileged information

tic;

%Construct omega_1 and omega_2

Omega_train_1 = kernel_matrix(P','lin_kernel');
Omega_train_2 = kernel_matrix(P',Kernel_class, Kernel_width);
Omega_train = Omega_train_1 + Omega_train_2;

%Construct tilde_omega_1 and tilde_omega_2

Omega_trainPF_1 = kernel_matrix(PF', 'lin_kernel',Kernel_width);
Omega_trainPF_2 = kernel_matrix(PF', Kernel_class, Kernel_width);
Omega_trainPF = Omega_trainPF_1+ Omega_trainPF_2;

%Calculate the output weights 

OutputWeight=(Omega_train+1/gamma*Omega_trainPF+eye(size(T,2))/C)\(T-C/gamma*ones(size(T))*Omega_trainPF)'; 
TrainingTime=toc;

%Calculate the training output
Y=(Omega_train * OutputWeight)'; 

%% Test using test data without privileged information

tic;
Omega_test_1 = kernel_matrix(P','lin_kernel',Kernel_width,TV.P');
Omega_test_2 = kernel_matrix(P',Kernel_class, Kernel_width,TV.P');
Omega_test = Omega_test_1 + Omega_test_2;
TY = (Omega_test' * OutputWeight)';                           
TestingTime=toc;
%% Calculate training and test accuracy

%regression case
if option.Type == 0
    TrainingAccuracy=sqrt(mse(T - Y));
    TestingAccuracy=sqrt(mse(TV.T - TY));           

%classification case
else
Y_temp=zeros(1, NumberofTrainingData);
Yt_temp=zeros(1,numel(TV.T));
for i=1:NumberofTrainingData
    [~,idx]=max(Y(:,i));
    Y_temp(i)=U_trainY(idx);
end
for i=1:numel(testY)
    [~,idx]=max(TY(:,i));
    Yt_temp(i)=U_trainY(idx);
end
TrainingAccuracy=length(find(Y_temp==trainY'))/NumberofTrainingData;

TestingAccuracy=length(find(Yt_temp==testY'))/NumberofTestingData;  
end


function omega = kernel_matrix(Xtrain,kernel_type, kernel_pars,Xt)

nb_data = size(Xtrain,1);


if strcmp(kernel_type,'RBF_kernel'),
    if nargin<4,
        XXh = sum(Xtrain.^2,2)*ones(1,nb_data);
        omega = XXh+XXh'-2*(Xtrain*Xtrain');
        omega = exp(-omega./kernel_pars(1));
    else
        XXh1 = sum(Xtrain.^2,2)*ones(1,size(Xt,1));
        XXh2 = sum(Xt.^2,2)*ones(1,nb_data);
        omega = XXh1+XXh2' - 2*Xtrain*Xt';
        omega = exp(-omega./kernel_pars(1));
    end
    
elseif strcmp(kernel_type,'lin_kernel')
    if nargin<4,
        omega = Xtrain*Xtrain';
    else
        omega = Xtrain*Xt';
    end
    
elseif strcmp(kernel_type,'poly_kernel')
    if nargin<4,
        omega = (Xtrain*Xtrain'+kernel_pars(1)).^kernel_pars(2);
    else
        omega = (Xtrain*Xt'+kernel_pars(1)).^kernel_pars(2);
    end
    
elseif strcmp(kernel_type,'wav_kernel')
    if nargin<4,
        XXh = sum(Xtrain.^2,2)*ones(1,nb_data);
        omega = XXh+XXh'-2*(Xtrain*Xtrain');
        
        XXh1 = sum(Xtrain,2)*ones(1,nb_data);
        omega1 = XXh1-XXh1';
        omega = cos(kernel_pars(3)*omega1./kernel_pars(2)).*exp(-omega./kernel_pars(1));
        
    else
        XXh1 = sum(Xtrain.^2,2)*ones(1,size(Xt,1));
        XXh2 = sum(Xt.^2,2)*ones(1,nb_data);
        omega = XXh1+XXh2' - 2*(Xtrain*Xt');
        
        XXh11 = sum(Xtrain,2)*ones(1,size(Xt,1));
        XXh22 = sum(Xt,2)*ones(1,nb_data);
        omega1 = XXh11-XXh22';
        
        omega = cos(kernel_pars(3)*omega1./kernel_pars(2)).*exp(-omega./kernel_pars(1));
    end
end